# KNN

- [K-近鄰演算法](https://zh.wikipedia.org/wiki/K-%E8%BF%91%E9%82%BB%E7%AE%97%E6%B3%95)
- [KNN visualization in just 13 lines of code](https://towardsdatascience.com/knn-visualization-in-just-13-lines-of-code-32820d72c6b6)

Step1: Import the required data and check the features.

In [15]:
from sklearn.datasets import load_iris

In [16]:
iris = load_iris()
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [13]:
# iris.target
# iris.target_names   
# iris.feature_names
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [14]:
# iris.data
iris.data.shape

(150, 4)

Step2: Split the data and Train the Model.

In [19]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=4)

NameError: name 'X' is not defined

In [39]:
import numpy as np
from sklearn.model_selection import train_test_split
X, y = np.arange(10).reshape((5, 2)), range(5)


X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.33, 
                                                    random_state=42)
X,y

(array([[0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]]),
 range(0, 5))

In [37]:
X_train,y_train

(array([[4, 5],
        [0, 1],
        [6, 7]]),
 [2, 0, 3])

In [38]:
X_test,y_test

(array([[2, 3],
        [8, 9]]),
 [1, 4])

## --- 老師方法----
從訓練資料中找出最近的K筆資料，然後看哪個類別的資料筆數最多

- 150 observations
- 4 features(sepal length, sepal width, petal length, petal width)
- Response variable is the iris species
- Classification problem since response is categorical.

In [20]:
import pandas as pd
import numpy as np
iris = pd.read_csv('data/iris.csv')
iris

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
..   ...            ...           ...            ...           ...   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   
149  150            5.9           3.0            5.1           1.8   

            Species  
0       Iris-setosa  
1       Iris-setosa  
2       Iris-setosa  
3       Iris-setosa  
4       Iris-setosa  
..              ...  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris-virginica  
149  Iris-virginica  

[150 rows x 6 columns]

In [21]:
# 找出與test最近的資料點
test = [5.0, 3.6, 1.4, 0.2]
s1 = (iris['SepalLengthCm'] - test[0])**2
s2 = (iris['SepalWidthCm'] - test[1])**2
s3 = (iris['PetalLengthCm'] - test[2])**2
s4 = (iris['PetalWidthCm'] - test[3])**2
d = (s1+s2+s3+s4)**0.5
iris.iloc[np.argmin(d)]

Id                         5
SepalLengthCm              5
SepalWidthCm             3.6
PetalLengthCm            1.4
PetalWidthCm             0.2
Species          Iris-setosa
Name: 4, dtype: object

In [ ]:
# 找出與test最近的K個資料點
K = 6
test = [5.0, 3.6, 1.4, 0.2]
s1 = (iris['SepalLengthCm'] - test[0])**2
s2 = (iris['SepalWidthCm'] - test[1])**2
s3 = (iris['PetalLengthCm'] - test[2])**2
s4 = (iris['PetalWidthCm'] - test[3])**2
d = (s1+s2+s3+s4)**0.5
iris.iloc[np.argsort(d)][0:K]

## KNN 視覺化

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn import datasets, neighbors
from mlxtend.plotting import plot_decision_regions

In [ ]:
def knn_comparison(data, k):
 x = data[['X','Y']].values
 y = data['class'].astype(int).values
 clf = neighbors.KNeighborsClassifier(n_neighbors=k)
 clf.fit(x, y)
# Plotting decision region
 plot_decision_regions(x, y, clf=clf, legend=2)
# Adding axes annotations
 plt.xlabel('X')
 plt.ylabel('Y')
 plt.title('Knn with K='+ str(k))
 plt.show()

### U Shaped

In [ ]:
data = pd.read_csv('演算法測試資料/ushape.csv')
for i in [1,5,20,30,40,80]:
    knn_comparison(data, i)

### 兩個類別資料互相被同心圓包圍 Two set concentric circles

In [ ]:
data = pd.read_csv('演算法測試資料/concertriccir2.csv')
for i in [1,5,20,30,40,80]:
    knn_comparison(data, i)

### XOR

In [ ]:
data = pd.read_csv('演算法測試資料/xor.csv')
for i in [1,5,20,30,40,80]:
    knn_comparison(data, i)

### Linearly separable

In [ ]:
data = pd.read_csv('演算法測試資料/linearsep.csv')
for i in [1,5,20,30,40,80]:
    knn_comparison(data, i)

### Outliers 有離異群的資料

In [ ]:
data = pd.read_csv('演算法測試資料/outlier.csv')
for i in [1,5,20,30,40,80]:
    knn_comparison(data, i)